crates/ruff_python_ast/src/scope.rs

```diff
@@ -14,12 +14,10 @@ pub struct Scope<'a> {
     /// A list of star imports in this scope. These represent _module_ imports (e.g., `sys` in
     /// `from sys import *`), rather than individual bindings (e.g., individual members in `sys`).
     star_imports: Vec<StarImportation<'a>>,
-    /// A map from bound name to binding index, for live bindings.
+    /// A map from bound name to binding index, for current bindings.
     bindings: FxHashMap<&'a str, BindingId>,
-    /// A map from bound name to binding index, for bindings that were created
-    /// in the scope but rebound (and thus overridden) later on in the same
-    /// scope.
-    pub rebounds: FxHashMap<&'a str, Vec<BindingId>>,
+    /// A map from bound name to binding index, for bindings that were shadowed later in the scope.
+    shadowed_bindings: FxHashMap<&'a str, Vec<BindingId>>,
```

Comment:
I wonder if using a flag `Vec` would be faster because most variables in a scope don't get shadowed. Scanning a small flat vec can be faster than searching  in a nested hash map (particularly if all entries fit into a single cache line, one entry here requires 12 bytes -> 5 entries fit into a cache line). 

Another solution could be to change `bindings` to `FxHashMap<&'a str, SmallVec<[BindingId; 2]>` . This has the benefit that we avoid allocating for almost all bindings (you have to double check if I did the math correctly but I think that `SmallVec<[BindingId;2]>` has the same size as `Vec<BindingId>`) and don't need another data structure for tracking.

Reply:
When you suggest using a Vec here for shadowed_bindings, do you mean something like Vec<(&'a str, BindingId)>?

Exactly (but without anonymous types :D)

Reply:
I'm going to leave as-is because we barely use these right now (only in a single rule) so it's hard for me to assess performance patterns in the real-world.